// n8n-nodes-globals/dist/credentials/GlobalConstantsCredentials.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GlobalConstantsCredentials = exports.GLOBAL_CONSTANTS_CREDENTIALS_NAME = void 0;
exports.GLOBAL_CONSTANTS_CREDENTIALS_NAME = 'globalConstantsApi';
class GlobalConstantsCredentials {
    constructor() {
        this.name = exports.GLOBAL_CONSTANTS_CREDENTIALS_NAME;
        this.displayName = 'Global Constants';
        this.properties = [
            {
                displayName: 'Format',
                name: 'format',
                type: 'options',
                options: [
                    {
                        name: 'Key-value pairs',
                        value: 'string',
                    },
                    {
                        name: 'JSON',
                        value: 'json',
                    },
                ],
                default: 'string',
                description: 'Choose the format for your global constants',
            },
            {
                displayName: 'Global Constants',
                name: 'globalConstants',
                type: 'string',
                default: '',
                placeholder: 'name1=value1\nname2=value2',
                hint: 'Use "name=value" format. Separate multiple constants with a new line.',
                typeOptions: {
                    rows: 10,
                },
                displayOptions: {
                    show: {
                        format: ['string'],
                    },
                },
            },
            {
                displayName: 'Global Constants',
                name: 'globalConstants',
                type: 'json',
                default: '{}',
                hint: 'Provide your constants as a JSON object. Example: { "obj": { "key": "value" }, "list": ["value1", "value2"] }',
                displayOptions: {
                    show: {
                        format: ['json'],
                    },
                },
            },
        ];
    }
}
exports.GlobalConstantsCredentials = GlobalConstantsCredentials;
