// n8n-nodes-instagram-token/dist/credentials/InstagramAccessTokenApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InstagramAccessTokenApi = void 0;
class InstagramAccessTokenApi {
    constructor() {
        this.name = 'instagramAccessTokenApi';
        this.displayName = 'Instagram Access Token API';
        this.documentationUrl = 'https://developers.facebook.com/docs/instagram-api/getting-started';
        this.properties = [
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
                description: 'Instagram access token (long-lived recommended)',
            },
            {
                displayName: 'Instagram Business Account ID',
                name: 'instagramBusinessAccountId',
                type: 'string',
                required: true,
                default: '',
                description: 'Instagram Business Account ID (auto-discovered if not provided)',
            },
            {
                displayName: 'Auto Refresh Token',
                name: 'autoRefresh',
                type: 'boolean',
                default: true,
                description: 'Automatically refresh token when expired (recommended)',
            },
            {
                displayName: 'Refresh Threshold (days)',
                name: 'refreshThreshold',
                type: 'number',
                default: 7,
                description: 'Refresh token when expiring within this many days',
                displayOptions: {
                    show: {
                        autoRefresh: [true],
                    },
                },
            },
            {
                displayName: 'Webhook Verify Token',
                name: 'webhookVerifyToken',
                type: 'string',
                required: false,
                default: '',
                description: 'Optional webhook verification token',
            },
        ];
        this.test = {
            request: {
                baseURL: 'https://graph.instagram.com',
                url: '/v23.0/me',
                qs: {
                    fields: 'id,username,name,account_type',
                    access_token: '={{$credentials.accessToken}}',
                },
            },
        };
    }
}
exports.InstagramAccessTokenApi = InstagramAccessTokenApi;
