// n8n-nodes-instagram-token/dist/nodes/Instagram/GenericFunctions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccessTokenAuto = getAccessTokenAuto;
exports.getInstagramBusinessAccountId = getInstagramBusinessAccountId;
exports.instagramApiRequest = instagramApiRequest;
exports.instagramApiRequestAllItems = instagramApiRequestAllItems;
exports.getTokenStatus = getTokenStatus;
exports.refreshAccessToken = refreshAccessToken;
exports.getAccessTokenHybrid = getAccessTokenHybrid;
exports.forceRefreshToken = forceRefreshToken;
exports.clearTokenCache = clearTokenCache;
exports.validateSignature = validateSignature;
const n8n_workflow_1 = require("n8n-workflow");
const tokenCache = new Map();
async function getAccessTokenAuto() {
    console.log('🔑 Using Instagram Access Token API (Hybrid System)');
    const credentials = await this.getCredentials('instagramAccessTokenApi');
    if (!credentials || !credentials.accessToken) {
        throw new Error('Instagram Access Token API credentials not found. Please configure your access token.');
    }
    return await getAccessTokenHybrid(this, 'default');
}
async function getInstagramBusinessAccountId() {
    try {
        const accessToken = await getAccessTokenAuto.call(this);
        const options = {
            method: 'GET',
            url: 'https://graph.instagram.com/v23.0/me',
            qs: {
                access_token: accessToken,
                fields: 'id,name,account_type,media_count,user_id,username',
            },
            json: true,
        };
        const response = await this.helpers.request(options);
        return response.id;
    }
    catch (error) {
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
async function instagramApiRequest(method, endpoint, body = {}, qs = {}, config = {}) {
    const accessToken = await getAccessTokenAuto.call(this);
    const { useAuthHeader = false } = config;
    console.log('🌐 Instagram API Request:');
    console.log('- Method:', method);
    console.log('- Endpoint:', endpoint);
    console.log('- Token (first 10 chars):', accessToken ? accessToken.substring(0, 10) + '...' : 'Missing');
    const options = {
        method,
        body,
        qs: {
            ...qs,
            ...(useAuthHeader ? {} : { access_token: accessToken }),
        },
        url: `https://graph.instagram.com/v23.0${endpoint}`,
        json: true,
    };
    if (useAuthHeader) {
        options.headers = {
            Authorization: `Bearer ${accessToken}`,
        };
    }
    try {
        console.log('🚀 Making request to Instagram API...');
        const response = await this.helpers.request(options);
        console.log('✅ Instagram API request successful');
        return response;
    }
    catch (error) {
        console.log('❌ Instagram API request failed:', error.message);
        throw new n8n_workflow_1.NodeApiError(this.getNode(), error);
    }
}
async function instagramApiRequestAllItems(method, endpoint, body = {}, qs = {}) {
    var _a;
    const returnData = [];
    let responseData;
    qs.limit = 100;
    do {
        responseData = await instagramApiRequest.call(this, method, endpoint, body, qs);
        returnData.push(...responseData.data);
    } while ((_a = responseData.paging) === null || _a === void 0 ? void 0 : _a.next);
    return returnData;
}
async function getTokenStatus(context, credentialId) {
    const credentials = await context.getCredentials('instagramAccessTokenApi');
    const token = credentials.accessToken;
    const cacheKey = `token_${credentialId}`;
    const cached = tokenCache.get(cacheKey);
    if (cached && cached.token === token && cached.expiresAt > Date.now() / 1000) {
        return {
            token: cached.token,
            isExpired: false,
            needsRefresh: false,
            expiresAt: cached.expiresAt,
            daysUntilExpiry: Math.ceil((cached.expiresAt - Date.now() / 1000) / (24 * 60 * 60)),
        };
    }
    try {
        const response = await context.helpers.request({
            method: 'GET',
            url: 'https://graph.instagram.com/debug_token',
            qs: {
                input_token: token,
                access_token: token,
            },
        });
        const data = response.data;
        const isExpired = data.is_valid === false;
        const needsRefresh = data.expires_at && (data.expires_at - Date.now() / 1000) < (7 * 24 * 60 * 60);
        tokenCache.set(cacheKey, {
            token,
            expiresAt: data.expires_at || 0,
            lastCheck: Date.now() / 1000,
        });
        return {
            token,
            isExpired,
            needsRefresh,
            expiresAt: data.expires_at || 0,
            daysUntilExpiry: data.expires_at ? Math.ceil((data.expires_at - Date.now() / 1000) / (24 * 60 * 60)) : 0,
        };
    }
    catch (error) {
        return {
            token,
            isExpired: false,
            needsRefresh: true,
            expiresAt: 0,
            daysUntilExpiry: 0,
        };
    }
}
async function refreshAccessToken(context, credentialId) {
    const credentials = await context.getCredentials('instagramAccessTokenApi');
    const token = credentials.accessToken;
    try {
        console.log('🔄 Attempting to refresh token...');
        const debugResponse = await context.helpers.request({
            method: 'GET',
            url: 'https://graph.instagram.com/debug_token',
            qs: {
                input_token: token,
                access_token: token,
            },
        });
        console.log('🔍 Token debug info:', JSON.stringify(debugResponse, null, 2));
        const response = await context.helpers.request({
            method: 'GET',
            url: 'https://graph.instagram.com/refresh_access_token',
            qs: {
                grant_type: 'ig_refresh_token',
                access_token: token,
            },
        });
        console.log('🔄 Refresh response:', JSON.stringify(response, null, 2));
        let responseData = response;
        if (typeof response === 'string') {
            console.log('🔄 Parsing JSON response...');
            responseData = JSON.parse(response);
        }
        const newToken = responseData.access_token;
        console.log('🔄 Parsed response data:');
        console.log('- access_token:', newToken ? newToken.substring(0, 10) + '...' : 'Missing');
        console.log('- expires_in:', responseData.expires_in);
        console.log('- Response keys:', Object.keys(responseData));
        if (!newToken) {
            console.log('❌ No access_token in refresh response');
            console.log('🔄 Response structure:', Object.keys(responseData));
            console.log('🔄 Response type:', typeof response);
            console.log('🔄 Full response:', JSON.stringify(responseData, null, 2));
            throw new Error('No access_token received from refresh endpoint');
        }
        const cacheKey = `token_${credentialId}`;
        tokenCache.set(cacheKey, {
            token: newToken,
            expiresAt: responseData.expires_in + Date.now() / 1000,
            lastCheck: Date.now() / 1000,
        });
        console.log(`🔑 New token generated. Please update your Instagram Access Token API credential with: ${newToken}`);
        console.log(`🔑 New token length: ${newToken.length}`);
        return newToken;
    }
    catch (error) {
        console.log('❌ Token refresh failed:', error.message);
        console.log('⚠️ This might be a long-lived token that cannot be refreshed automatically');
        console.log('⚠️ Please manually update your Instagram Access Token API credential with a new token');
        console.log('🔄 Using original token as fallback');
        return token;
    }
}
async function getAccessTokenHybrid(context, credentialId) {
    const credentials = await context.getCredentials('instagramAccessTokenApi');
    const token = credentials.accessToken;
    console.log('🔑 Instagram Access Token Debug:');
    console.log('- Token (first 10 chars):', token ? token.substring(0, 10) + '...' : 'Missing');
    console.log('- Auto Refresh:', credentials.autoRefresh);
    console.log('- Full token length:', token ? token.length : 0);
    if (credentials.autoRefresh === false) {
        console.log('🔑 Using token directly (auto-refresh disabled)');
        return token;
    }
    console.log('🔍 Checking token status...');
    const status = await getTokenStatus(context, credentialId);
    console.log('🔍 Token Status:');
    console.log('- Is Expired:', status.isExpired);
    console.log('- Needs Refresh:', status.needsRefresh);
    console.log('- Days Until Expiry:', status.daysUntilExpiry);
    console.log('- Status Token (first 10 chars):', status.token ? status.token.substring(0, 10) + '...' : 'Missing');
    if (status.isExpired || status.needsRefresh) {
        console.log('🔄 Token needs refresh, refreshing automatically...');
        const newToken = await refreshAccessToken(context, credentialId);
        console.log('🔄 New token received (first 10 chars):', newToken ? newToken.substring(0, 10) + '...' : 'Missing');
        return newToken;
    }
    console.log('✅ Using current token');
    console.log('✅ Final token (first 10 chars):', status.token ? status.token.substring(0, 10) + '...' : 'Missing');
    return status.token;
}
async function forceRefreshToken(context, credentialId) {
    try {
        const newToken = await refreshAccessToken(context, credentialId);
        return {
            success: true,
            newToken,
            message: 'Token refreshed successfully',
        };
    }
    catch (error) {
        return {
            success: false,
            newToken: '',
            message: `Failed to refresh token: ${error.message}`,
        };
    }
}
function clearTokenCache(credentialId) {
    const cacheKey = `token_${credentialId}`;
    tokenCache.delete(cacheKey);
}
function validateSignature(body, signature, secret) {
    if (!secret || typeof secret !== 'string' || secret.trim() === '') {
        console.log('⚠️ Webhook secret not configured or invalid, skipping signature validation');
        return true;
    }
    if (!signature || typeof signature !== 'string') {
        console.log('⚠️ Webhook signature not provided, skipping validation');
        return true;
    }
    try {
        const crypto = require('crypto');
        const expectedSignature = crypto
            .createHmac('sha256', secret)
            .update(body)
            .digest('hex');
        return crypto.timingSafeEqual(Buffer.from(signature, 'hex'), Buffer.from(expectedSignature, 'hex'));
    }
    catch (error) {
        console.log('⚠️ Error validating webhook signature:', error.message);
        return true;
    }
}
